pkg/blobstore/grpcclients/cas_zstd_blob_access.go

```diff
@@ -0,0 +1,265 @@
+	resourceName string
+	writeOffset  int64
+	cancel       context.CancelFunc
+}
+
+func (w *zstdByteStreamWriter) Write(p []byte) (int, error) {
+	if err := w.client.Send(&bytestream.WriteRequest{
+		ResourceName: w.resourceName,
+		WriteOffset:  w.writeOffset,
+		Data:         p,
+	}); err != nil {
+		return 0, err
+	}
+	w.writeOffset += int64(len(p))
+	w.resourceName = ""
+	return len(p), nil
+}
+
+func (w *zstdByteStreamWriter) Close() error {
+	if err := w.client.Send(&bytestream.WriteRequest{
+		ResourceName: w.resourceName,
+		WriteOffset:  w.writeOffset,
+		FinishWrite:  true,
+	}); err != nil {
+		w.cancel()
+		w.client.CloseAndRecv()
+		return err
+	}
+	_, err := w.client.CloseAndRecv()
+	w.cancel()
+	return err
+}
+
+func (ba *casWithZstdBlobAccess) Put(ctx context.Context, digest digest.Digest, b buffer.Buffer) error {
+	if digest.GetSizeBytes() < ba.compressionThresholdBytes {
+		return ba.uncompressedBlobAccess.Put(ctx, digest, b)
+	}
+
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+	resourceName := digest.GetByteStreamWritePath(uuid.Must(ba.uuidGenerator()), remoteexecution.Compressor_ZSTD)
+	client, err := ba.byteStreamClient.Write(
+		metadata.AppendToOutgoingContext(ctxWithCancel, resourceNameHeader, resourceName),
+	)
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	byteStreamWriter := &zstdByteStreamWriter{
+		client:       client,
+		resourceName: resourceName,
+		writeOffset:  0,
+		cancel:       cancel,
+	}
+
+	zstdWriter, err := zstd.NewWriter(byteStreamWriter, zstd.WithEncoderConcurrency(1))
+	if err != nil {
+		cancel()
+		client.CloseAndRecv()
+		return status.Errorf(codes.Internal, "Failed to create zstd writer: %v", err)
+	}
+
+	if err := b.IntoWriter(zstdWriter); err != nil {
+		zstdWriter.Close()
+		byteStreamWriter.Close()
+		return err
+	}
+
+	if err := zstdWriter.Close(); err != nil {
+		byteStreamWriter.Close()
+		return err
+	}
+
+	return byteStreamWriter.Close()
+}
+
+func (ba *casWithZstdBlobAccess) FindMissing(ctx context.Context, digests digest.Set) (digest.Set, error) {
+	return ba.uncompressedBlobAccess.FindMissing(ctx, digests)
+}
+
+func (ba *casWithZstdBlobAccess) GetCapabilities(ctx context.Context, instanceName digest.InstanceName) (*remoteexecution.ServerCapabilities, error) {
+	cacheCapabilities, err := getCacheCapabilities(ctx, ba.capabilitiesClient, instanceName)
+	if err != nil {
+		return nil, err
+	}
+
+	if !slices.Contains(cacheCapabilities.SupportedCompressors, remoteexecution.Compressor_ZSTD) {
+		return nil, status.Error(codes.FailedPrecondition, "Server does not support ZSTD compression")
+	}
+
+	// Only return fields that pertain to the Content Addressable
+	// Storage. Include compression support information.
+	return &remoteexecution.ServerCapabilities{
+		CacheCapabilities: &remoteexecution.CacheCapabilities{
+			DigestFunctions:      digest.RemoveUnsupportedDigestFunctions(cacheCapabilities.DigestFunctions),
+			SupportedCompressors: cacheCapabilities.SupportedCompressors,
```

Comment:
I don't think that propagating `supported_compressors` here makes sense. Doing so would cause these compressors to be merged with what's specified in bb_storage's configuration.